### ISSUE TITLE
`flutter_scene` failing to build on `main`

### ISSUE THREAD

**Issue by domesticmouse** [bdero/flutter_scene#48]
From https://github.com/flutter/codelabs/actions/runs/12263106435/job/34213915579?pr=2270

```
Exception: Failed to run flutter_scene_importer:import command in build hook (exit code 255):
STDERR: Building assets for package:flutter_scene_importer failed.
build_output.json contained a format error.

Contents: {"version":"1.5.0","timestamp":"2024-12-10 19:36:03.000","dependencies":[]}.
Only compatible versions with 1.6.0 are supported (was: 1.5.0).

Error: Compiling native assets failed.

STDOUT: 
#0      buildModels (package:flutter_scene_importer/build_hooks.dart:46:7)
#1      main.<anonymous closure> (file:///Users/runner/work/codelabs/codelabs/intro_flutter_gpu/step_12/hook/build.dart:10:5)
#2      build (package:native_assets_cli/src/api/build.dart:101:16)
#3      main (file:///Users/runner/work/codelabs/codelabs/intro_flutter_gpu/step_12/hook/build.dart:9:3)
#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:313:33)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:194:12)
```

**Comment by bdero** [comment]
I was able to reproduce this on the example app & step 12 of the codelab against framework commit 953208752e. Commit eef8daf0 is known working, so things broke somewhere between framework commit eef8daf0 and 953208752e.

**Comment by bdero** [comment]
Started failing on:
`0ca9f51227 Roll Flutter Engine from 7dd1cffaad06 to a8c08bc9b909 (5 revisions) (#159990)`

**Comment by bdero** [comment]
That engine roll has two Dart SDK rolls:
- https://github.com/flutter/engine/pull/57066
- https://github.com/flutter/engine/pull/57060

https://github.com/flutter/engine/pull/57066 contains a [dependency update](https://dart.googlesource.com/sdk.git/+/71f7a42f4219ce3864834df76b9c644e92c15075%5E%21/#F0) for https://dart.googlesource.com/native.git:
028f89d4a59448f9ecda5bd3d4519e2d7c742072 -> 8ea1a9db0af42933eb22334c4506ca464d7237e9

`0caab92e` in native.git seems to bump a resource version in `native_assets_cli` under `model/build_config` from 1.5.0 to 1.6.0:
```
0caab92e [native_assets_cli] Rename `supportedAssetTypes` to `buildAssetTypes` (#1786)
```

**Comment by bdero** [comment]
@dcharkes It looks like our buildhooks broke after [0caab92e](https://dart.googlesource.com/native.git/+/0caab92e7edba7fea24ca82e7861a7daa829c2ea%5E%21/#F28) was rolled into Flutter engine/framework. How should we update the buildhooks or dependencies to keep them working? Does a new version of `native_assets_cli` need to be pushed to pub?

**Comment by dcharkes** [comment]
> `0caab92e` in native.git seems to have made a version check for a resource more restrictive in `native_assets_cli` under `model/build_config`: https://dart.googlesource.com/native.git/+/0caab92e7edba7fea24ca82e7861a7daa829c2ea%5E%21/#F28

I see, the version checks were broken during a previous refactoring. I fixed the check to allow version skew in the build config. But apparently I missed the check for allowing version skew in the build output. And the version skew integration tests were also deleted in said refactoring unfortunately.

The dependencies are rolled manually in `flutter_tools`, but the command failing here is one from the Dart executable I presume?

We've got two ways to fix it:

1. Support version skew again properly in the implementation. And roll that. (We need to do this anyway, otherwise it will break on the next version number bump again.)
2. Release new versions of the dart-lang/native packages that speak 1.6.0.

> Does a new version of `native_assets_cli` need to be pushed to pub?

For option 2, yes.

I can take a look at that tomorrow to unblock you.

**Comment by dcharkes** [comment]
> * Support version skew again properly in the implementation. And roll that. (We need to do this anyway, otherwise it will break on the next version number bump again.)

Done. But we'll have to wait until Dart rolls into Flutter to see things go green via this route.

> * Release new versions of the dart-lang/native packages that speak 1.6.0.

Done. I've put up a PR for the first in the stack of packages:

* https://github.com/bdero/flutter_gpu_shaders/pull/5

**Comment by kevinzhow** [comment]
@domesticmouse `Flutter 3.28.0-0.1.pre` is using 1.5.0, so I modified the build_hooks.dart as workaround
1. use fvm install the `3.28.0-0.1.pre` verson
2.modify the `build_hooks.dart` in flutter_scene
```dart
    final importerResult = Process.runSync(
      "<your_local_path>/fvm/versions/3.28.0-0.1.pre/bin/dart",
```

**Comment by domesticmouse** [comment]
> @domesticmouse `Flutter 3.28.0-0.1.pre` is using 1.5.0, so I modified the build_hooks.dart as workaround
> 
> 1. use fvm install the `3.28.0-0.1.pre` verson
> 2. modify the `build_hooks.dart` in flutter_scene
> 
> ```dart
>     final importerResult = Process.runSync(
>       "<your_local_path>/fvm/versions/3.28.0-0.1.pre/bin/dart",
> ```

The `flutter/codelabs` repo doesn't use `fvm`. I have CI checks for `stable`, `beta` and `master` channels. At this point, the codelab compiles with none of the channels. I'm not going to pin CI to a version, as that becomes an ongoing maintenance pain. 

I'm leaving this open until we have a fix that works on `master` so I can re-enable the Flutter codelabs CI for the Intro to Flutter GPU codelab.

**Comment by domesticmouse** [comment]
For context, here's what I'm seeing attempting to build the `flutter_scene` codelab step [`step_12`](https://github.com/flutter/codelabs/tree/main/intro_flutter_gpu/step_12)

```console
$ flutter run          
Launching lib/main.dart on iPhone 16 in debug mode...
Running Xcode build...                                                  
Xcode build done.                                            4.8s
Failed to build iOS app
Error (Xcode): Unhandled exception:


Could not build the application for the simulator.
Error launching application on iPhone 16.
```

And of course, `flutter doctor`:

```console
flutter doctor -v
[✓] Flutter (Channel main, 3.28.0-2.0.pre.38579, on macOS 15.2 24C101 darwin-arm64, locale
    en-AU)
    • Flutter version 3.28.0-2.0.pre.38579 on channel main at /Users/brett/flutter
    • Upstream repository https://github.com/flutter/flutter
    • Framework revision 09a585ba1d (12 days ago), 2024-12-26 16:06:41 -0800
    • Engine revision 09a585ba1d
    • Dart version 3.7.0 (build 3.7.0-267.0.dev)
    • DevTools version 2.41.0

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/brett/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
      This is the JDK bundled with the latest Android Studio installation on this machine.
      To manually set the JDK path, use: `flutter config --jdk-dir="path/to/jdk"`.
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16C5032a
    • CocoaPods version 1.16.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)

[✓] IntelliJ IDEA Community Edition (version 2024.3.1.1)
    • IntelliJ at /Users/brett/Applications/IntelliJ IDEA Community Edition.app
    • Flutter plugin version 83.0.4
    • Dart plugin version 243.23177

[✓] VS Code (version 1.96.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.102.0

[✓] Connected device (5 available)
    • Brett’s iPad (mobile)           • 00008103-001169D42223001E            • ios            • iOS 18.2 22C152
    • iPhone 16 (mobile)              • A1852855-5DE4-4C89-855F-9001C87A4FBC • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-18-2 (simulator)
    • macOS (desktop)                 • macos                                • darwin-arm64   • macOS 15.2 24C101 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad                • darwin         • macOS 15.2 24C101 darwin-arm64
    • Chrome (web)                    • chrome                               • web-javascript • Google Chrome 131.0.6778.264
    ! Error: Browsing on the local area network for iPhone. Ensure the device is unlocked and attached with a cable or associated
      with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

**Comment by domesticmouse** [comment]
This is now working on head for macOS and Linux. Windows is failing to build in CI. I'll do some investigation and open a new issue once I have a clue what is breaking.
